src/sinks/util/unix.rs

```diff
@@ -197,12 +241,88 @@ where
 
         Ok(())
     }
+
+    // Same as UdpSink
+    async fn run_datagram(self: Box<Self>, input: BoxStream<'_, Event>) -> Result<(), ()> {
+        let bytes_sent = register!(BytesSent::from(Protocol::UNIX));
+        let mut input = input.peekable();
+
+        let mut encoder = self.encoder.clone();
+        while Pin::new(&mut input).peek().await.is_some() {
+            let mut socket = match self.connector.connect_backoff().await {
+                UnixEither::Datagram(datagram) => datagram,
+                UnixEither::Stream(_) => {
+                    unreachable!("run_datagram is only called with Datagram mode")
+                }
+            };
+
+            while let Some(mut event) = input.next().await {
+                let byte_size = event.estimated_json_encoded_size_of();
+
+                self.transformer.transform(&mut event);
+
+                let finalizers = event.take_finalizers();
+                let mut bytes = BytesMut::new();
+
+                // Errors are handled by `Encoder`.
+                if encoder.encode(event, &mut bytes).is_err() {
+                    continue;
+                }
+
+                match udp_send(&mut socket, &bytes).await {
+                    Ok(()) => {
+                        emit!(SocketEventsSent {
+                            mode: SocketMode::Unix,
+                            count: 1,
+                            byte_size,
+                        });
+
+                        bytes_sent.emit(ByteSize(bytes.len()));
+                        finalizers.update_status(EventStatus::Delivered);
+                    }
+                    Err(error) => {
+                        emit!(UnixSocketSendError {
+                            path: &self.connector.path,
+                            error: &error
+                        });
+                        finalizers.update_status(EventStatus::Errored);
+                        break;
+                    }
+                }
+            }
+        }
```

Comment:
I wonder if we can reuse this code instead of duplicating it from `udp.rs`.

Reply:
It would certainly be nice. Something like a plain function taking a bunch of parameters (`bytes_sent`, `input`, `encoder`, `socket`, `transformer` and the socket's path in the case of UDS), or could this somehow be done with a trait(?) given how `encoder` and `transformer` are accessed via `self` in both UDP and UDS?
To be honest, I don't really work with Rust so I'm not entirely familiar with how things get done there.

Reply:
I think just a free function with some parameters would do it.